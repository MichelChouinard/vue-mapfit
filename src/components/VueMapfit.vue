<template>
  <div id="vue-mapfit"></div>
</template>

<script>
export default {
  props: {
    apikey: {
      type: String,
    },
    center: {
      type: [Array, Object],
      required: true,
    },
    zoom: {
      type: Number,
      default: 16,
    },
  },

  methods: {
    initMapfit() {
      const { mapfit } = window;

      if (this.apikey) mapfit.apikey = this.apikey;

      // draw map
      const map = mapfit.MapView('vue-mapfit', { theme: 'day' });

      const position = mapfit.LatLng(this.center);
      const marker = mapfit.Marker(position);

      // set the map center on marker position
      map.setCenter(position);

      // add marker to map
      map.addMarker(marker);

      // set map zoom
      map.setZoom(this.zoom);
    },

    createScriptTag() {
      const mfScript = document.createElement('script');
      mfScript.setAttribute('type', 'text/javascript');
      mfScript.setAttribute('defer', '');
      mfScript.setAttribute('src', 'http://cdn.mapfit.com/v2-4/assets/js/mapfit.js');

      return mfScript;
    },

    createStyleTag() {
      const mfStyle = document.createElement('link');
      mfStyle.setAttribute('rel', 'stylesheet');
      mfStyle.setAttribute('href', 'http://cdn.mapfit.com/v2-4/assets/css/mapfit.css');

      return mfStyle;
    },
  },

  mounted() {
    if (typeof window === 'undefined') return;

    if (typeof window.mapfit === 'undefined') {
      const style = this.createStyleTag();
      const script = this.createScriptTag();

      script.addEventListener('load', () => this.initMapfit());

      document.head.appendChild(style);
      document.body.appendChild(script);
    } else {
      this.initMapfit();
    }
  },
};
</script>
