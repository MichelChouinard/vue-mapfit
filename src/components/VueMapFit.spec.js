import { mount } from '@vue/test-utils';
import VueMapfit from 'src/components/VueMapfit';

let wrapper;

const center = {
  lat: 47.850361,
  lng: 12.057812,
};

describe('VueMapfit display', () => {
  beforeEach(() => {
    wrapper = mount(VueMapfit, {
      propsData: {
        center,
      },
    });
  });

  test('matches snapshot using defaults', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe('VueMapfit methods generate correct scripts', () => {
  beforeEach(() => {
    wrapper = mount(VueMapfit, {
      propsData: {
        center,
      },
    });
  });

  test('creates correct script tag', () => {
    const mfScript = wrapper.vm.createScriptTag(); // returns HTMLScriptElement

    expect(mfScript.type).toBe('text/javascript');
    expect(mfScript.defer).toBe(true);
    expect(mfScript.src).toBe('http://cdn.mapfit.com/v2-4/assets/js/mapfit.js');
  });

  test('creates correct script tag', () => {
    const mfStyle = wrapper.vm.createStyleTag(); // returns HTMLLinkElement

    expect(mfStyle.rel).toBe('stylesheet');
    expect(mfStyle.href).toBe('http://cdn.mapfit.com/v2-4/assets/css/mapfit.css');
  });
});

describe('VueMapfit initialization', () => {
  global.document.head.appendChild = jest.fn();
  global.document.body.appendChild = jest.fn();

  beforeEach(() => {
    wrapper = mount(VueMapfit, {
      propsData: {
        center,
      },
    });
  });

  test('required scripts and styles are already available in browser', () => {
    const styleTag = document.createElement('link');
    styleTag.setAttribute('rel', 'stylesheet');
    styleTag.setAttribute('href', 'http://cdn.mapfit.com/v2-4/assets/css/mapfit.css');

    const scriptTag = document.createElement('script');
    scriptTag.setAttribute('type', 'text/javascript');
    scriptTag.setAttribute('defer', '');
    scriptTag.setAttribute('src', 'http://cdn.mapfit.com/v2-4/assets/js/mapfit.js');

    expect(global.document.head.appendChild).toHaveBeenCalledWith(styleTag);
    expect(global.document.body.appendChild).toHaveBeenCalledWith(scriptTag);
  });
});
