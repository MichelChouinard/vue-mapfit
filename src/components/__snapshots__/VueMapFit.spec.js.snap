// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueMapfit display matches snapshot using defaults 1`] = `
<div
  id="vue-mapfit"
/>
`;
